async_producer.go

```diff
@@ -1227,6 +1231,19 @@ func (p *asyncProducer) retryHandler() {
 		}
 
 		buf.Add(msg)
+
+		if maxBufferSize > 0 && buf.Length() >= maxBufferSize {
+			msgToHandle := buf.Peek().(*ProducerMessage)
```

Comment:
Is there not a race condition here? In that, we merely `Peek()` and then operate based on what we peeked, but the `buf.Remove` is not guaranteed to be removing the same `*ProducerMessage` that we just peeked?

Reply:
Thank you for raising this concern! In this implementation, the buf queue is exclusively accessed within the retryHandler function, which runs in a single goroutine. Since there is no concurrent access to buf, operations like Peek() and Remove() are safe and sequential, ensuring the correct message is handled.

Additionally, the same pattern of using Peek() followed by Remove() is already used in lines 1223 and 1224 of this file. https://github.com/IBM/sarama/blob/c1c3fa4c761722df6914c953cdc1bd3f05b74b23/async_producer.go#L1223 This consistency reinforces that the operations are expected to be thread-safe in the current single-threaded context.

Please correct me if i'm wrong.

Reply:
👍 Sounds good. My first thoughts are always concern about concurrency. But if `buf` is scope limited to this one goroutine, then you’re right. There’s no concerns about concurrency, as it’s exclusively held.